invisible_cities/cities/detsim_simulate_electrons.py

```diff
@@ -0,0 +1,52 @@
+import numpy  as np
+
+from typing    import   List, Tuple, Callable
+
+#######################################
+######### ELECTRON SIMULATION #########
+#######################################
+def generate_electrons(energies    : np.array,
+                       wi          : float,
+                       fano_factor : float ) -> np.array:
+    """ generate secondary electrons from energy deposits
+    """
+    nes  = np.array(energies/wi, dtype = int)
+    pois = nes < 10
+    nes[ pois] = np.random.poisson(nes[pois])
+    nes[~pois] = np.round(np.random.normal(nes[~pois], np.sqrt(nes[~pois] * fano_factor)))
+    return nes
+
+
+def drift_electrons(zs             : np.array,
+                    electrons      : np.array,
+                    lifetime       : float,
+                    drift_velocity : float) -> np.array:
+    """ returns number of electrons due to lifetime loses from secondary electrons
+    """
+    ts  = zs / drift_velocity
+    nes = electrons - np.random.poisson(electrons * (1. - np.exp(-ts/lifetime)))
+    nes[nes < 0] = 0
+    return nes
+
+
+def diffuse_electrons(xs                     : np.array,
+                      ys                     : np.array,
+                      zs                     : np.array,
+                      electrons              : np.array,
+                      transversal_diffusion  : float,
+                      longitudinal_diffusion : float)\
+                      -> Tuple[np.array, np.array, np.array, np.array]:
+    """
+    starting from hits with positions xs, ys, zs, and number of electrons,
+    apply diffusion and return diffused positions xs, ys, zs for each electron
+    """
+    xs = np.repeat(xs, electrons.astype(int))
+    ys = np.repeat(ys, electrons.astype(int))
+    zs = np.repeat(zs, electrons.astype(int))
+
+    sqrtz = zs ** 0.5
+    dxs  = np.random.normal(xs, sqrtz * transversal_diffusion)
+    dys  = np.random.normal(ys, sqrtz * transversal_diffusion)
+    dzs  = np.random.normal(zs, sqrtz * longitudinal_diffusion)
+
+    return (dxs, dys, dzs)
```

Comment:
If I fully understand the code, right now we are assuming that every ionization e- is being created at the same time.
This is always true except for Bi214 simulations. The decay of Bi214 lead to hits "delayed" tens or hundreds of microseconds with respect the first energy depositions. This time difference must be considered when diffusing electrons in order to properly reconstruct their Z positions and therefore, the detectors response.

Reply:
In fact I can think of other types of event where this is relevant too (neutron captures for example). Things like this are included in full sim but we'll have to be careful to have this functionality here too. Things like this were one of the to-dos left hanging in the old versions

Reply:
That's right, but if you look at dataflow PR #700, you see that the part that manages the buffer times is different from the one from electron simulation. I think that it would be more natural to include this shift in buffer times directly there, since here we just implement the effects of attachment and diffusion, which is independent of relative times between hits. Nevertheless, we could add this as a PR once detsim is approved.

Reply:
We're not talking about buffer time but absolute time. The MC/hit has the time (in PR #694 you don't even return the info which should be sorted) as well as position, in the BiPo events Javi mentioned the hits could be at ver similar x, y, z position but separated by multiple µs or even ms.

Reply:
Yes, we are saying the same. Sure, PR #694, must have this info too. What I say is that if you look at PR #700, the "buffer time" part also computes the drift times (`z/drif_velocity`). So to take into account the absolute times (or relative times between hit at initial time 0), this is were this should be implemented, by summing up the drift time plus the hit production time.

Reply:
I don't think I agree, especially if we want to be as flexible as possible but, anyway, let's see how things progress with the other functions